bundles/org.openhab.core.thing/src/test/java/org/openhab/core/thing/internal/ThingManagerImplTest.java

```diff
@@ -17,103 +17,129 @@
+import org.openhab.core.config.core.ConfigDescriptionRegistry;
+import org.openhab.core.config.core.validation.ConfigDescriptionValidator;
+import org.openhab.core.events.EventPublisher;
 import org.openhab.core.service.ReadyService;
 import org.openhab.core.storage.Storage;
 import org.openhab.core.storage.StorageService;
 import org.openhab.core.thing.Thing;
 import org.openhab.core.thing.ThingUID;
 import org.openhab.core.thing.binding.ThingHandlerFactory;
+import org.openhab.core.thing.i18n.ThingStatusInfoI18nLocalizationService;
+import org.openhab.core.thing.internal.ThingTracker.ThingTrackerEvent;
+import org.openhab.core.thing.link.ItemChannelLinkRegistry;
+import org.openhab.core.thing.type.ChannelGroupTypeRegistry;
+import org.openhab.core.thing.type.ChannelTypeRegistry;
+import org.openhab.core.thing.type.ThingTypeRegistry;
 import org.openhab.core.util.BundleResolver;
 import org.osgi.framework.Bundle;
-import org.osgi.service.component.ComponentContext;
 
 /**
  * @author Simon Kaufmann - Initial contribution
  */
+@NonNullByDefault
 public class ThingManagerImplTest {
 
-    private @Mock BundleResolver mockBundleResolver;
-    private @Mock Bundle mockBundle;
-    private @Mock ComponentContext mockComponentContext;
-    private @Mock ReadyService mockReadyService;
-    private @Mock Thing mockThing;
-
-    private @Mock StorageService mockStorageService;
-    private @Mock Storage<Object> mockStorage;
-
-    private final ThingRegistryImpl thingRegistry = new ThingRegistryImpl();
+    private @Mock @NonNullByDefault({}) Bundle bundleMock;
+    private @Mock @NonNullByDefault({}) BundleResolver bundleResolverMock;
+    private @Mock @NonNullByDefault({}) ChannelGroupTypeRegistry channelGroupTypeRegistryMock;
+    private @Mock @NonNullByDefault({}) ChannelTypeRegistry channelTypeRegistryMock;
+    private @Mock @NonNullByDefault({}) CommunicationManager communicationManagerMock;
+    private @Mock @NonNullByDefault({}) EventPublisher eventPublisherMock;
+    private @Mock @NonNullByDefault({}) ConfigDescriptionRegistry configDescriptionRegistryMock;
+    private @Mock @NonNullByDefault({}) ConfigDescriptionValidator configDescriptionValidatorMock;
+    private @Mock @NonNullByDefault({}) ItemChannelLinkRegistry itemChannelLinkRegistryMock;
+    private @Mock @NonNullByDefault({}) ThingTypeRegistry thingTypeRegistryMock;
+    private @Mock @NonNullByDefault({}) ReadyService readyServiceMock;
+    private @Mock @NonNullByDefault({}) SafeCaller safeCallerMock;
+    private @Mock @NonNullByDefault({}) Storage<Object> storageMock;
+    private @Mock @NonNullByDefault({}) StorageService storageServiceMock;
+    private @Mock @NonNullByDefault({}) Thing thingMock;
+    private @Mock @NonNullByDefault({}) ThingRegistryImpl thingRegistryMock;
+
+    // This class is final so it cannot be mocked
+    private final ThingStatusInfoI18nLocalizationService thingStatusInfoI18nLocalizationService = new ThingStatusInfoI18nLocalizationService();
```

Comment:
There are 57 classes defined as "public final class".
If they cannot be easily mocked in tests, it will either result in custom build Mocks or less tests.

Reply:
I think the root cause with this particular case is that the class doesn't implement an interface which would otherwise be perfectly mockable. :wink: